### DISCUSSION TITLE
Any completely 100% zero allocation / no heap escape / no garbage generated compression for golang?

### DISCUSSION THREAD

**Discussion by kolinfluence** [klauspost/compress#808] [Q&A] [answered]
knowing u are the expert in this area, do u know of anything like this? i dont mind the compression ratio, even if 10% lesser is fine. or smaller for lesser than 4mb (which seemed to not escape to heap that much) or something similar?

anything u know of or can provide a clue on how to get this done? will be compressing and decompressing lots of such data and would prefer not having any memory allocation for garbage collection.

just anything. appreciate this help. thx.
p.s. : your packages are used everywhere so thanks a lot. will buy u a coffee if u have sponsorship page.

**Comment by klauspost** [comment, marked as answer]
[S2 blocks](https://pkg.go.dev/github.com/klauspost/compress/s2#Encode) only allocate on the stack.

S2 streams with [FlushOnWrite](https://pkg.go.dev/github.com/klauspost/compress/s2#WriterFlushOnWrite) and "WriterConcurrency(1)" should also be fully allocation-less. Of course you must reuse the Writers you allocate.

**Comment by klauspost** [comment]
zstd will also be mostly allocationless when using [EncodeAll](https://pkg.go.dev/github.com/klauspost/compress@v1.16.5/zstd#Encoder.EncodeAll) and providing properly sized buffers, once the internal buffers have been allocated.

Using [WithLowerEncoderMem(false)](https://pkg.go.dev/github.com/klauspost/compress@v1.16.5/zstd#WithLowerEncoderMem) will make it allocate buffers that are unlikely to need re-allocation.
